### DISCUSSION TITLE
GSoC 2026: Implementing RFDiffusion in DeepChem

### DISCUSSION THREAD

**Discussion by chidwipak** [deepchem/deepchem#5107] [Show and tell]
# RFdiffusion in DeepChem — GSoC 2026

Hello everyone, I’m Chidwipak, a final-year Computer Science student at IIIT Sri City, specializing in AI/ML. I’ve mainly been working on AI for scientific computing, drug discovery, and distributed systems, and I already had some background in drug discovery before starting this project.

## How I started

I joined the DeepChem community around December 2025. I started by going through the documentation, tutorials, contribution guide, and the codebase to understand how everything is organized. Since I already had some experience with scientific ML and drug discovery, it was quite easy for me to understand the problem side. The DeepChem part was new, so I spent some time getting used to how models, datasets, testing, documentation, and CI are handled here.

When I saw the RFdiffusion project, it felt like a good fit for me because I had already worked with molecular generation and molecular representations, and I was familiar with AlphaFold2, RoseTTAFold, and related protein structure work.

Before opening the main PRs, I first wanted to understand the whole RFdiffusion model properly. So I spent around a month reading the papers, going through the original RFdiffusion repository and the RoseTTAFold code, and looking at the relevant AlphaFold2 parts. I also built a complete prototype locally first. I wanted to see the whole thing working end to end before trying to bring it into DeepChem.

Once that was working, I discussed the approach with Bharath and started splitting the work into smaller PRs. In the first few PRs, I also had to get used to DeepChem's style, especially docstrings, examples, formatting, tests, CI, and how to structure stacked PRs. After that, it became much more straightforward.

The overall plan was simple. DeepChem didn't have the diffusion foundation needed for RFdiffusion, so I first built that part and made sure the basic model flow worked. Then I moved to the actual RFdiffusion architecture and split that into smaller modules.

```text
ProteinBackboneFeaturizer (#4970)
              |
              v
Diffusion foundation (#4994 - #4997)
              |
              v
RFdiffusion geometry: frames, IPA, pair track, sequence/structure track
              |
              v
SO(3) diffusion + motif conditioning
              |
              v
Final RFdiffusion integration (#5102)
              |
              v
RFdiffusion All-Atom (#5103 - #5106) [ongoing]
```

The first RFdiffusion PR was [#4970](https://github.com/deepchem/deepchem/pull/4970), the `ProteinBackboneFeaturizer`. It reads the protein structure and extracts the N, CA, and C backbone coordinates for each residue. That gave the rest of the model a standard input to work with.

After that came the diffusion foundation. [#4994](https://github.com/deepchem/deepchem/pull/4994) added the embedding layers, [#4995](https://github.com/deepchem/deepchem/pull/4995) added the cosine diffusion schedule, [#4996](https://github.com/deepchem/deepchem/pull/4996) added the simple Transformer denoiser, and [#4997](https://github.com/deepchem/deepchem/pull/4997) wrapped everything into `RFDiffusionModel`.

I kept the Transformer simple on purpose. I wanted to first check that the basic diffusion pipeline itself was working: adding noise, removing noise, training, sampling, and saving and loading the model. Once that part was working, I could move to the actual RFdiffusion architecture.

Then I started building the geometric part.

[#4982](https://github.com/deepchem/deepchem/pull/4982) adds the rigid-frame operations. This is where each residue gets a position and orientation instead of just being treated as a flat coordinate vector. It also includes the frame construction, composition, inversion, and rotation exp/log operations.

After that, [#5028](https://github.com/deepchem/deepchem/pull/5028) adds Invariant Point Attention. This lets the model work with the 3D geometry while staying consistent if the whole structure is rotated or translated.

[#5036](https://github.com/deepchem/deepchem/pull/5036) adds the pair track. It keeps information about the relationship between residue pairs, which is important because a lot of protein structure information is about how residues relate to each other.

Then [#5041](https://github.com/deepchem/deepchem/pull/5041) brings the residue features, pair features, and frame updates together into the multi-track network. This is where the RoseTTAFold-style part of the model started to come together.

In parallel, [#5073](https://github.com/deepchem/deepchem/pull/5073) adds SO(3) rotational diffusion. The cosine schedule handles translation, while SO(3) handles rotation, so together they let the model diffuse the full residue frames.

[#5076](https://github.com/deepchem/deepchem/pull/5076) adds contig parsing and motif conditioning. This is the part that lets us keep a fixed motif and generate new backbone around it.

All of these modules were tested separately, but that still wasn't the full model. They had to be connected.

That is what [#5102](https://github.com/deepchem/deepchem/pull/5102) does. It connects the real RFdiffusion multi-track network to `RFDiffusionModel`, connects the translation and SO(3) diffusion parts, and adds the motif-conditioning path. The earlier Transformer is still available as the simpler option, but the real RFdiffusion network is now the main architecture.

While doing that final integration, I also found two problems in the SO(3) part. One was causing NaNs during backpropagation, and another was making the rotational diffusion setup much slower than it should be. I fixed both as part of the integration.

For testing, I didn't rely only on whether the code runs. The individual modules have unit tests and doctests, and the math-heavy parts were checked against the original references where possible. For example, the cosine schedule was checked against Improved DDPM, and the SO(3) implementation was checked against the RFdiffusion reference. I also tested the full model flow with training, generation, save/reload, and the DeepChem CI checks.

So that is basically the RFdiffusion part of the project. The main backbone implementation is now in place, and the final integration is in [#5102](https://github.com/deepchem/deepchem/pull/5102).

## What I am working on next

After finishing the backbone, I started working on the RFdiffusion All-Atom extension as additional work.

I am following the same idea here too. Instead of putting everything into one large PR, I am splitting it into smaller pieces that can be checked separately.

The current draft PRs are:

- [#5103](https://github.com/deepchem/deepchem/pull/5103) — ligand parsing and all-atom loss functions
- [#5104](https://github.com/deepchem/deepchem/pull/5104) — symmetry and conditioning
- [#5105](https://github.com/deepchem/deepchem/pull/5105) — protein structure quality metrics
- [#5106](https://github.com/deepchem/deepchem/pull/5106) — RFdiffusion All-Atom model integration

These are still ongoing draft PRs. The idea is to build the supporting pieces first and then bring them together into the full All-Atom model.

The RFdiffusion work is useful for DeepChem because it brings protein backbone generation into the same ecosystem as the other DeepChem models and datasets. The All-Atom extension is the next step toward more detailed protein-ligand and structure-conditioned design.

I plan to keep working on the All-Atom side and continue contributing to DeepChem after GSoC as well.

Thanks to Jose Antonio Siguenza for all the reviews and feedback on the PRs, and to Bharath Ramsundar for the opportunity and guidance throughout the project.

**Comment by bobby404-K** [comment]
Incredible writeup and massive congratulations on the project, Chidwipak! 🚀

RFdiffusion is a huge milestone for DeepChem’s generative toolkit. I really appreciate how methodical your roadmap was—building out the diffusion primitives and unit-testing the math-heavy components against reference implementations before wiring up the full multi-track network.

Super excited to see the All-Atom PRs progressing as well. Great job sticking through the tough debugging and delivering such a clean foundation for the community!
